### PR TITLE
Add new BeExactly assertions for DateTimeOffset

### PR DESCRIPTION
**Pull request by gliljas**
Not quite ready yet. More a basis for discussion.

The value formatting was one issue. I don't think the offset should be hidden, even if it's zero. It's no less significant than a positive or negative offset.

The assertion could perhaps be made with a Then step and a "...but the offsets differ" message.

### CODE REVIEW

File: Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs [hunk 1]
```diff
@@ -21,6 +21,7 @@ public bool CanHandle(object value)
         public void Format(object value, FormattedObjectGraph formattedGraph, FormattingContext context, FormatChild formatChild)
         {
             DateTimeOffset dateTimeOffset;
+            bool significantOffset = false;
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs:24]: Can you try to move those changes in a separate commit so we can easily see why this change was needed?

File: Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs [hunk 2]
```diff
@@ -70,12 +72,15 @@ public void Format(object value, FormattedObjectGraph formattedGraph, Formatting
...
-            if (dateTimeOffset.Offset < TimeSpan.Zero)
+            else if (dateTimeOffset.Offset < TimeSpan.Zero)
             {
                 formattedGraph.AddFragment(" ");
                 formatChild("offset", dateTimeOffset.Offset, formattedGraph);
             }
+            else if (significantOffset && (hasDate || hasTime))
+            {
+                formattedGraph.AddFragment(" UTC");
+            }
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs:83]: 🤔 What happens if the last `else if` is not happening?

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs [hunk 3]
```diff
@@ -68,7 +68,7 @@ public DateTimeOffsetAssertions(DateTimeOffset? value)
         }
 
         /// <summary>
-        /// Asserts that the current <see cref="DateTimeOffset"/> is exactly equal to the <paramref name="expected"/> value.
+        /// Asserts that the current <see cref="DateTimeOffset"/> represents the same point in time as the <paramref name="expected"/> value.
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:71]: 🤔 Shouldn't we also update the failure message to cover this important distinction?

File: Tests/FluentAssertions.Specs/Primitives/DateTimeOffsetAssertionSpecs.cs [hunk 4]
```diff
@@ -364,6 +364,189 @@ public void
 
         #endregion
 
+        #region (Not) BeExactly
+
+        [Fact]
+        public void Should_succeed_when_asserting_datetimeoffset_value_is_exactly_equal_to_the_same_value()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Primitives/DateTimeOffsetAssertionSpecs.cs:370]: 🔧 No need to mention `datetimeoffset` in there since it's clear from the context. Also, we prefer to keep literal names of code elements out of the test name.

File: Tests/FluentAssertions.Specs/Primitives/DateTimeOffsetAssertionSpecs.cs [hunk 5]
```diff
@@ -364,6 +364,189 @@ public void
...
+        {
+            // Arrange
+            DateTimeOffset dateTime = new DateTime(2016, 06, 04);
+            DateTimeOffset sameDateTime = new DateTime(2016, 06, 04);
+
+            // Act
+            Action act = () => dateTime.Should().BeExactly(sameDateTime);
+
+            // Assert
+            act.Should().NotThrow();
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Primitives/DateTimeOffsetAssertionSpecs.cs:380]: 🔧 If the assertion is not expected to throw anything, you don't need to explicitly use `NotThrow` (anymore).

File: Tests/FluentAssertions.Specs/Primitives/DateTimeOffsetAssertionSpecs.cs (hunk 4 above)
Comment (gliljas) [reply Tests/FluentAssertions.Specs/Primitives/DateTimeOffsetAssertionSpecs.cs:370]: Gladly. I just copied the Be/NotBe tests, which has the misleading "numeric" name, and chanhed it. Will fix.

File: Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs (hunk 2 above)
Comment (gliljas) [reply Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs:83]: That would leave us in the original path. No offset appended/prepended.

File: Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs (hunk 1 above)
Comment (gliljas) [reply Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs:24]: The formatter changes? I'll do anything you ask for, as soon as I understand what it is :)

My rationale for the formatting change was:

1. Without the offset mentioned, even if it's TimeSpan.Zero, the assertion messages can become cryptic. "Exptected 2021-06-01 +1h, but it was 2021-06-01"
2. I'm just guessing, but I think it's actually more common to deal with DateTimeOffsets which have a non-zero offset, so the change shouldn't be that big of a deal.
3. The offset is skipped for plain DateTime. I thought that would make sense.

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs (hunk 3 above)
Comment (gliljas) [reply Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:71]: Done. However, the messages became a bit wordy. Perhaps too much so.

File: Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs (hunk 1 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs:24]: > My rationale for the formatting change was:

@jnyrup what do you think? Makes sense to me.

> The formatter changes? I'll do anything you ask for, as soon as I understand what it is :)

To split your changes into two commits. One with the formatter changes (with the rational in the title and description), and one with the new `DateTimeOffset` assertions. See also https://www.continuousimprover.com/2020/03/keep-source-control-history-clean.html

File: Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs (hunk 2 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs:83]: I would add an `else` with a comment explaining that. Avoids unnecessary frowns later on.

File: Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs (hunk 1 above)
Comment (gliljas) [reply Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs:24]: Also, using "UTC" as the offset was just a choice that seemed OK at the time. UTC is not an offset, it's a time zone. There are other time zones which have a zero offset, so something like "Z", "+0h" or even "±0h" could make more sense.

Comment (dennisdoomen) [reply Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs:24]: I think `+0h` would be clearer.

Comment (gliljas) [reply Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs:24]: Agreed. Done. I will fix the commits according to you liking if and when we conclude the matter. The documentation will need some additions too.

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs (hunk 3 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:71]: Maybe, but it'll make it very clear why the test failed.

File: Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs [hunk 6]
```diff
@@ -70,12 +72,19 @@ public void Format(object value, FormattedObjectGraph formattedGraph, Formatting
                 formattedGraph.AddFragment(" +");
                 formatChild("offset", dateTimeOffset.Offset, formattedGraph);
             }
-
-            if (dateTimeOffset.Offset < TimeSpan.Zero)
+            else if (dateTimeOffset.Offset < TimeSpan.Zero)
             {
                 formattedGraph.AddFragment(" ");
```
Comment (jnyrup) [inline Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs:77]: Shouldn't this have a minus sign?
```suggestion
                formattedGraph.AddFragment(" -");
```

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs [hunk 7]
```diff
@@ -59,7 +59,117 @@ public DateTimeOffsetAssertions(DateTimeOffset? value)
             params object[] becauseArgs)
         {
             Execute.Assertion
+                .BecauseOf(because, becauseArgs)
+                .WithExpectation("Expected {context:the date and time} to represent the same point in time as {0}{reason}, ", expected)
+                .ForCondition(Subject.HasValue)
+                .FailWith("but found a <null> DateTimeOffset.")
+                .Then
                 .ForCondition(Subject == expected)
+                .FailWith("but its value {0} doesn't.", Subject)
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:68]:
```suggestion
                .FailWith("but {0} does not.", Subject)
```

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs [hunk 8]
```diff
@@ -101,11 +211,11 @@ public DateTimeOffsetAssertions(DateTimeOffset? value)
         /// <param name="becauseArgs">
         /// Zero or more objects to format using the placeholders in <paramref name="because" />.
         /// </param>
-        public AndConstraint<TAssertions> NotBe(DateTimeOffset unexpected, string because = "",
+        public AndConstraint<TAssertions> NotBeExactly(DateTimeOffset unexpected, string because = "",
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:214]: Should the summary be updated?

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs [hunk 9]
```diff
@@ -78,11 +188,11 @@ public DateTimeOffsetAssertions(DateTimeOffset? value)
         /// <param name="becauseArgs">
         /// Zero or more objects to format using the placeholders in <paramref name="because" />.
         /// </param>
-        public AndConstraint<TAssertions> Be(DateTimeOffset? expected, string because = "",
+        public AndConstraint<TAssertions> BeExactly(DateTimeOffset? expected, string because = "",
             params object[] becauseArgs)
         {
             Execute.Assertion
-                .ForCondition(Subject == expected)
+                .ForCondition((Subject == null && expected == null) || (Subject != null && expected != null && Subject.Value.EqualsExact(expected.Value)))
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:195]: We strive to keep lines at atmost 130 characters wide.
```suggestion
                .ForCondition((Subject == null && expected == null) 
                    || (Subject != null && expected != null && Subject.Value.EqualsExact(expected.Value)))
```

File: Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs (hunk 6 above)
Comment (gliljas) [reply Src/FluentAssertions/Formatting/DateTimeOffsetValueFormatter.cs:77]: I was confused by that one too, but indeed, it's added by the formatter.

Yes, there's test coverage.

**Review by dennisdoomen — APPROVED** [review]

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs [hunk 10]
```diff
@@ -80,18 +85,55 @@ public DateTimeOffsetAssertions(DateTimeOffset? value)
...
+            else
+            {
+                Execute.Assertion
+                    .BecauseOf(because, becauseArgs)
+                    .WithExpectation("Expected {context:the date and time} to represent the same point in time as {0}{reason}, ", expected)
+                    .ForCondition(Subject.HasValue)
+                    .FailWith("but found a <null> DateTimeOffset.", expected)
+                    .Then
+                    .ForCondition(Subject == expected)
+                    .FailWith("but {1} does not.", expected, Subject)
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:105]:
```suggestion
                    .FailWith("but {0} does not.", Subject)
```

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs [hunk 11]
```diff
@@ -101,12 +143,102 @@ public DateTimeOffsetAssertions(DateTimeOffset? value)
...
+        public AndConstraint<TAssertions> BeExactly(DateTimeOffset expected, string because = "",
+            params object[] becauseArgs)
+        {
+            Execute.Assertion
+                .BecauseOf(because, becauseArgs)
+                .WithExpectation("Expected {context:the date and time} to be exactly {0}{reason}, ", expected)
+                .ForCondition(Subject.HasValue)
+                .FailWith("but found a <null> DateTimeOffset.")
+                .Then
+                .ForCondition(Subject.Value.EqualsExact(expected) == true)
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:177]: I haven't checked the code out, but I would believe `== true` to be superfluous.
```suggestion
                .ForCondition(Subject.Value.EqualsExact(expected))
```

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs [hunk 12]
```diff
@@ -101,12 +143,102 @@ public DateTimeOffsetAssertions(DateTimeOffset? value)
...
+            else
+            {
+                Execute.Assertion
+                    .BecauseOf(because, becauseArgs)
+                    .WithExpectation("Expected {context:the date and time} to be exactly {0}{reason}, ", expected)
+                    .ForCondition(Subject.HasValue)
+                    .FailWith("but found a <null> DateTimeOffset.", expected)
+                    .Then
+                    .ForCondition(Subject.Value.EqualsExact(expected.Value))
+                    .FailWith("but it was {0}.", expected, Subject)
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:216]:
```suggestion
                    .FailWith("but it was {0}.", Subject)
```

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs [hunk 13]
```diff
@@ -101,12 +143,102 @@ public DateTimeOffsetAssertions(DateTimeOffset? value)
...
+                   .ForCondition(!Subject.HasValue)
+                   .FailWith("Expected {context:the date and time} to be <null>{reason}, but it was {0}.", Subject);
+            }
+            else
+            {
+                Execute.Assertion
+                    .BecauseOf(because, becauseArgs)
+                    .WithExpectation("Expected {context:the date and time} to be exactly {0}{reason}, ", expected)
+                    .ForCondition(Subject.HasValue)
+                    .FailWith("but found a <null> DateTimeOffset.", expected)
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:213]:
```suggestion
                    .FailWith("but found a <null> DateTimeOffset.")
```

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs [hunk 14]
```diff
@@ -80,18 +85,55 @@ public DateTimeOffsetAssertions(DateTimeOffset? value)
...
+                   .ForCondition(!Subject.HasValue)
+                   .FailWith("Expected {context:the date and time} to be <null>{reason}, but it was {0}.", Subject);
+            }
+            else
+            {
+                Execute.Assertion
+                    .BecauseOf(because, becauseArgs)
+                    .WithExpectation("Expected {context:the date and time} to represent the same point in time as {0}{reason}, ", expected)
+                    .ForCondition(Subject.HasValue)
+                    .FailWith("but found a <null> DateTimeOffset.", expected)
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:102]:
```suggestion
                    .FailWith("but found a <null> DateTimeOffset.")
```

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs [hunk 15]
```diff
@@ -101,12 +143,102 @@ public DateTimeOffsetAssertions(DateTimeOffset? value)
...
+        /// <param name="becauseArgs">
+        /// Zero or more objects to format using the placeholders in <paramref name="because" />.
+        /// </param>
+        public AndConstraint<TAssertions> NotBeExactly(DateTimeOffset unexpected, string because = "",
+            params object[] becauseArgs)
+        {
+            Execute.Assertion
+                .ForCondition(Subject?.EqualsExact(unexpected) != true)
+                .BecauseOf(because, becauseArgs)
                 .FailWith("Did not expect {context:the date and time} to be {0}{reason}, but it was.", unexpected);
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:242]:
```suggestion
                .FailWith("Did not expect {context:the date and time} to be exactly {0}{reason}, but it was.", unexpected);
```

File: Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs [hunk 16]
```diff
@@ -123,11 +255,11 @@ public DateTimeOffsetAssertions(DateTimeOffset? value)
...
         /// </param>
-        public AndConstraint<TAssertions> NotBe(DateTimeOffset? unexpected, string because = "",
+        public AndConstraint<TAssertions> NotBeExactly(DateTimeOffset? unexpected, string because = "",
             params object[] becauseArgs)
         {
             Execute.Assertion
-                .ForCondition(Subject != unexpected)
+                .ForCondition(!((Subject == null && unexpected == null) || (Subject != null && unexpected != null && Subject.Value.EqualsExact(unexpected.Value))))
                 .BecauseOf(because, becauseArgs)
                 .FailWith("Did not expect {context:the date and time} to be {0}{reason}, but it was.", unexpected);
```
Comment (jnyrup) [inline Src/FluentAssertions/Primitives/DateTimeOffsetAssertions.cs:264]:
```suggestion
                .FailWith("Did not expect {context:the date and time} to be exactly {0}{reason}, but it was.", unexpected);
```

**Review by jnyrup — CHANGES_REQUESTED** [review]
Aside from the failing tests, two other step are necessary to finish up this PR
* Run `AcceptChanges.ps1/sh` to update the accepted public methods.
* Add a line to the release notes.

Both steps are explained in https://github.com/fluentassertions/fluentassertions/blob/release-6.0/CONTRIBUTING.md, but ping me if anything is unclear or causing trouble.

**Review by jnyrup — APPROVED** [review]
